redis/cluster.py

```diff
@@ -0,0 +1,1249 @@
+            1. A ``db`` querystring option, e.g. redis://localhost?db=0
+            2. If using the redis:// or rediss:// schemes, the path argument
+               of the url, e.g. redis://localhost/0
+            3. A ``db`` keyword argument to this function.
+
+        If none of these options are specified, the default db=0 is used.
+
+        All querystring options are cast to their appropriate Python types.
+        Boolean arguments can be specified with string values "True"/"False"
+        or "Yes"/"No". Values that cannot be properly cast cause a
+        ``ValueError`` to be raised. Once parsed, the querystring arguments
+        and keyword arguments are passed to the ``ConnectionPool``'s
+        class initializer. In the case of conflicting arguments, querystring
+        arguments always win.
+
+        """
+        return cls(url=url, **kwargs)
+
+    def on_connect(self, connection):
+        """
+        Initialize the connection, authenticate and select a database and send
+         READONLY if it is set during object initialization.
+        """
+        connection.set_parser_class(ClusterParser)
+        connection.on_connect()
+
+        if self.read_from_replicas:
+            # Sending READONLY command to server to configure connection as
+            # readonly. Since each cluster node may change its server type due
+            # to a failover, we should establish a READONLY connection
+            # regardless of the server type. If this is a primary connection,
+            # READONLY would not affect executing write commands.
+            connection.send_command('READONLY')
+            if str_if_bytes(connection.read_response()) != 'OK':
+                raise ConnectionError('READONLY command failed')
+
+    def get_redis_connection(self, node):
+        if not node.redis_connection:
+            self.nodes_manager.create_redis_connections([node])
+        return node.redis_connection
+
+    def get_node(self, node_name=None, host=None, port=None):
+        return self.nodes_manager.get_node(node_name, host, port)
+
+    def get_all_primaries(self):
+        return self.nodes_manager.get_nodes_by_server_type(PRIMARY)
+
+    def get_all_replicas(self):
+        return self.nodes_manager.get_nodes_by_server_type(REPLICA)
+
+    def get_random_node(self):
+        return random.choice(list(self.nodes_manager.nodes_cache.values()))
+
+    def get_all_nodes(self):
+        return list(self.nodes_manager.nodes_cache.values())
+
+    def pubsub(self, node=None, host=None, port=None, **kwargs):
+        """
+        Allows passing a ClusterNode, or host&port, to get a pubsub instance
+        connected to the specified node
+        """
+        return ClusterPubSub(self, node=node, host=host, port=port, **kwargs)
+
+    def _determine_nodes(self, *args, **kwargs):
+        command = args[0]
+        node_flag = self.nodes_flags.get(command)
+
+        if node_flag == RANDOM:
+            return [self.get_random_node()]
```

Comment:
Changes to call the above, renamed functions
...

Reply:
changed